### DISCUSSION TITLE
actuation for multilinked thrust-actuated robot

### DISCUSSION THREAD

**Discussion by sugikazu75** [loco-3d/crocoddyl#1292] [Ideas]
Thank you for developing and maintaining this repository. 

I want to apply this library to a multi-link robot with thrusters on each links. 
The case where the thruster is fixed to the root link is supported by `include/crocoddyl/multibody/actuations/floating-base-thrusters.hpp`. 
However, when the positions of the thrusters change with the movement of joints, I think the actuation class should have an instance of pinocchio::Data, but I wonder if there is a better way. 
I'm not sure, so please let me know if there is a better way based on the design philosophy of this library. 
I think it has something to do with what is mentioned in https://github.com/loco-3d/crocoddyl/pull/1213#issue-2096612933 and https://github.com/loco-3d/crocoddyl/issues/1239.

**Comment by cmastalli** [comment]
Hi @sugikazu75,

As you figured out, Crocoddyl does not yet support thrusters in limbs. To nicely exploit the recursive nature of RNEA or ABA algorithms, we should pass the thruster forces to those algorithms, rather than computing ourselves the Jacobians. However, our current API is designed to implement the latter solution.

In short, I don't think we need to introduce pinocchio::Data for getting the Jacobians, but rather store this information in a sort of Crocoddyl's collector and adapt the calls of the RNEA and ABA algorithms.

We can work on this collaboratively as it would be useful to have this type of actuation model in Crocoddyl. A more efficient way of work could be through meetings where we discuss API changes before implementing them.
